src/Microsoft.Health.Dicom.Core/Features/CustomTag/CustomTagService.cs

```diff
@@ -18,17 +23,70 @@ public class CustomTagService : ICustomTagService
+                { DicomVR.DS.Code,  CustomTagIndexType.StringIndex },
+                { DicomVR.DT.Code,  CustomTagIndexType.DateTimeIndex },
+                { DicomVR.FL.Code,  CustomTagIndexType.DoubleIndex },
+                { DicomVR.FD.Code,  CustomTagIndexType.DoubleIndex },
+                { DicomVR.IS.Code,   CustomTagIndexType.StringIndex },
+                { DicomVR.LO.Code,   CustomTagIndexType.StringIndex },
+                { DicomVR.PN.Code,   CustomTagIndexType.StringIndex },
+                { DicomVR.SH.Code,  CustomTagIndexType.StringIndex },
+                { DicomVR.SL.Code,   CustomTagIndexType.LongIndex },
+                { DicomVR.SS.Code,   CustomTagIndexType.LongIndex },
+                { DicomVR.TM.Code,   CustomTagIndexType.DateTimeIndex },
+                { DicomVR.UI.Code,   CustomTagIndexType.StringIndex },
+                { DicomVR.UL.Code,   CustomTagIndexType.LongIndex },
+                { DicomVR.US.Code,   CustomTagIndexType.LongIndex },
+            };
+
+        /// <summary>
+        /// Mapping from IndexType to the delete function
+        /// </summary>
+        private readonly IReadOnlyDictionary<CustomTagIndexType, Func<string, int, CancellationToken, Task<long>>> indexTypeAndDeleteFunctionMapping;
+
+        /// <summary>
+        /// Max records are deleted in each transaction.
+        /// </summary>
+        private const int MaxDeleteRecordCount = 10000;
+
+        /// <summary>
+        /// Delay between each delete transaction.
+        /// </summary>
+        private const int DelayBetweenDeleteTransaction = 100;
+
+        public CustomTagService(ICustomTagStore customTagStore, ICustomTagEntryValidator customTagEntryValidator, IDicomTagParser dicomTagParser, ILogger<CustomTagService> logger)
         {
             EnsureArg.IsNotNull(customTagStore, nameof(customTagStore));
             EnsureArg.IsNotNull(customTagEntryValidator, nameof(customTagEntryValidator));
+            EnsureArg.IsNotNull(dicomTagParser, nameof(dicomTagParser));
             EnsureArg.IsNotNull(logger, nameof(logger));
 
             _customTagStore = customTagStore;
             _customTagEntryValidator = customTagEntryValidator;
+            _dicomTagParser = dicomTagParser;
             _logger = logger;
+
+            indexTypeAndDeleteFunctionMapping = new Dictionary<CustomTagIndexType, Func<string, int, CancellationToken, Task<long>>>()
+            {
+                { CustomTagIndexType.StringIndex, _customTagStore.DeleteCustomTagStringIndexAsync },
+                { CustomTagIndexType.LongIndex, _customTagStore.DeleteCustomTagLongIndexAsync },
+                { CustomTagIndexType.DoubleIndex, _customTagStore.DeleteCustomTagDoubleIndexAsync },
+                { CustomTagIndexType.DateTimeIndex, _customTagStore.DeleteCustomTagDateTimeIndexAsync },
+                { CustomTagIndexType.PersonNameIndex, _customTagStore.DeleteCustomTagPersonNameIndexAsync },
+            };
```

Comment:
Would it make more sense to simply have one method on the store for deletion? Storing in different tables is sort of a SQL-specific implementation detail, right? So would it make sense for the SQL custom tag store to change the table based on the tag's type?

Reply:
I think that also make sense.
What my code comes from is to keep consistent with that:   XXXStore is simply a wrapper on storage procedure (e.g :https://github.com/microsoft/dicom-server/blob/main/src/Microsoft.Health.Dicom.SqlServer/Features/Store/SqlIndexDataStore.cs), more generic logic would be on the Core layer.

Reply:
this is solved by simplify deletion logic